### PR TITLE
[5.6] Add Http2ServerPush Middleware

### PR DESCRIPTION
**Pull request by JacobBennett**
Allowing users to enable Http2 Server Push easily by adding this middleware to the `Http/Kernel.php` web middleware stack.

Would be accompanied by a PR to `laravel/laravel` if merged.

Know there are a few concerns that @GrahamCampbell had about this one and would welcome feedback on what we could do to improve this before merge. Thanks!

### CODE REVIEW

**Comment by browner12** [comment]
hey @JacobBennett , this looks awesome. Was hoping for an easy way to integrate server push into my sites.

Could you add a ELI5 comment about what process this takes?

From what I can gather from the code, it looks like it takes the response content, and extracts all the assets out of there.  Then it builds up a list of those assets, generates the appropriate server push headers, and attaches them to the response.  Is that the gist?

Is there a way to prevent certain assets from being included in this, or do you get all or nothing?

excited to have this!

**Comment by JacobBennett** [comment]
Hey @browner12, think you got the gist. Crawls the response looking for 'pushable' assets and then generates the proper `link` headers to make that happen.

Could certainly create some configs around this middleware such as.

- types of assets to include (scripts, styles, images)
- types of assets to exclude
- limit on the number of items to push (this is already available as an argument)

Currently though, it is all or nothing.

**Comment by JacobBennett** [comment]
Referencing [this article](https://www.smashingmagazine.com/2017/04/guide-http2-server-push/#final-thoughts), Would like to add the following to this middleware.

- [ ] Ability to limit types of assets pushed (css, js, images)
- [ ]  Ability to track returning visitors and prevent push headers

File: src/Illuminate/Http/Middleware/Http2ServerPush.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+     * @param \Illuminate\Http\Request $request
+     * @param \Closure $next
+     *
+     * @return mixed
+     */
+    public function handle(Request $request, Closure $next, $limit = null)
+    {
+        $response = $next($request);
+
+        if ($response->isRedirection() || ! $response instanceof Response || $request->isJson()) {
```
Comment (sisve) [inline src/Illuminate/Http/Middleware/Http2ServerPush.php]: Does it really matter if `$request->isJson()`? I imagine that it's important what the response is, not what the request that generated the response contains.

Comment (sisve) [reply src/Illuminate/Http/Middleware/Http2ServerPush.php]: Also, the order of checks. Should you check for the Response type before calling `->isRedirection()`?

File: src/Illuminate/Http/Middleware/Http2ServerPush.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            '.JPEG' => 'image',
+            '.PNG'  => 'image',
+            '.TIFF' => 'image',
+        ];
+
+        $type = collect($linkTypeMap)->first(function ($type, $extension) use ($url) {
+            return str_contains(strtoupper($url), $extension);
+        });
+
+        return is_null($type) ? null : "<{$url}>; rel=preload; as={$type}";
```
Comment (sisve) [inline src/Illuminate/Http/Middleware/Http2ServerPush.php]: Does this need any escaping? Are we sure that $url is always in a proper format and never contains any `>` or other special characters?

File: src/Illuminate/Http/Middleware/Http2ServerPush.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            '.BMP'  => 'image',
+            '.GIF'  => 'image',
+            '.JPG'  => 'image',
+            '.JPEG' => 'image',
+            '.PNG'  => 'image',
+            '.TIFF' => 'image',
+        ];
+
+        $type = collect($linkTypeMap)->first(function ($type, $extension) use ($url) {
+            return str_contains(strtoupper($url), $extension);
```
Comment (sisve) [inline src/Illuminate/Http/Middleware/Http2ServerPush.php]: Shouldn't this be a "if url ends with"? You should probably be parsing the $url and just check the path segment for matches, ignoring things like the host, query and fragment.

File: src/Illuminate/Http/Middleware/Http2ServerPush.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+     *
+     * @param \Illuminate\Http\Response $response
+     *
+     * @return \Illuminate\Support\Collection
+     */
+    protected function fetchLinkableNodes($response)
+    {
+        $crawler = $this->getCrawler($response);
+
+        return collect($crawler->filter('link, script[src], img[src]')->extract(['src', 'href']));
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Middleware/Http2ServerPush.php]: The problem is that this picks up on everything, and only things from the same exact domain can be pushed. Not just anything in such tags.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by GrahamCampbell** [comment]
I honestly think this doesn't belong in the core tbh.

**Comment by GrahamCampbell** [comment]
It makes a fine package, and should stay as a package.

**Comment by JacobBennett** [comment]
@taylorotwell want to answer your question first. Essentially, server push provides similar benefits to inlining all of your css in your page but without the negatives to that approach, namely the inability to cache assets.

the `Link` headers specify assets that are present on the page and allow them to be downloaded in the first response that the browser is returned from the server.  This has the end affect of decreasing page load times.

image

As can be seen in the image above, if using HTTP2 push to just push through the css, you can can faster load times than inlining all css with a HTTP1 call. This will also be true for First paint time and DOMContentLoaded time.

Currently this middleware pushes all images, `js`, and `css` it finds on the page. I would like to add a configuration option on the middleware that pushes `css` by default and allows for selective enabling of the others. 

This [HTTP2 Server Push Article](https://www.smashingmagazine.com/2017/04/guide-http2-server-push) is a great reference for learning about what server push is and what it allows you to do.

File: src/Illuminate/Http/Middleware/Http2ServerPush.php (hunk 4 above)
Comment (JacobBennett) [reply src/Illuminate/Http/Middleware/Http2ServerPush.php]: I think we could filter this list using the  `config('app.url')` to prevent pushing items from other domains. Thoughts?

Comment (sisve) [reply src/Illuminate/Http/Middleware/Http2ServerPush.php]: @JacobBennett That sounds wrong. The config entry is for url generation from artisan command. You could instead use the host from the $request instance.

File: src/Illuminate/Http/Middleware/Http2ServerPush.php (hunk 1 above)
Comment (JacobBennett) [reply src/Illuminate/Http/Middleware/Http2ServerPush.php]: Great points here. Re-ordered the checks, and now checking the response for JSON instead of the `request`

**Comment by taylorotwell** [comment]
Using the Crawler component to track down that stuff doesn't offset the peformance gains?

**Comment by HDVinnie** [comment]
I personally use your package and think it is a great middleware and does heed performance gains. It would be neat to have this built in the core of laravel since it is for performance purposes. However it should be configurable on what to "push". I had to modify the middleware to not do images. Myself and many others get 502 bad gateway on pages with many images.

**Comment by GrahamCampbell** [comment]
The other thing going through my mind, is that server push doesn't have to be handled by the app. There's no reason why a higher up layer can't add these headers, such as the load balancer.

**Comment by HDVinnie** [comment]
@GrahamCampbell couldn't the same be said for trusted proxy that was added to laravel?

**Comment by GrahamCampbell** [comment]
> couldn't the same be said for trusted proxy that was added to laravel?

Yeh, but this middleware is defintely more oppinionated.

**Comment by JacobBennett** [comment]
@taylorotwell to be fair I haven't tested how the crawler component affects the speed of the responses. Might not be a bad benchmark to run. There may be more efficient ways to locate pushable resources using straight regex as well. Is this a metric you would like some hard numbers on before considering this PR?

@GrahamCampbell I agree that this could be implemented higher up the stack, referencing the previous article however:

image

@HDVinnie Thanks for the support! Glad to hear the package is working well for you. My goal is to make it more configurable regardless of if it exists in core or not.

**Comment by X-Coder264** [comment]
@JacobBennett nginx 1.13.9 (which was released a month ago) added HTTP/2 server push support 😃 

https://nginx.org/en/CHANGES

``` 
Changes with nginx 1.13.9                                        20 Feb 2018

    *) Feature: HTTP/2 server push support; the "http2_push" and
       "http2_push_preload" directives.
```

**Comment by taylorotwell** [comment]
I wonder if a faster / alternative approach would be using a `preload` helper which is similar to what Symfony did: https://symfony.com/blog/new-in-symfony-3-3-asset-preloading-with-http-2-push

That way we're not having to use the Crawler component or a regular expression potentially. The preloads are explicitly defined and we can just retrieve them directly.

**Comment by JacobBennett** [comment]
@taylorotwell love that idea, and actually, that is really similar to what @tomschlick middleware does. Maybe we could adapt his instead of adapting mine. 

@tomschlick I could do the integration into core if you want or you could open a new PR yourself so you get credit for the work you did on the package? Happy to help in any way I can.

**Comment by tomschlick** [comment]
@JacobBennett Yeah I'm game for that, I'll try to get something worked up later today.

**Comment by JacobBennett** [comment]
@tomschlick did you want to put something together or would you like me to create a first iteration on it based on your package and give you credit? Would likely just be doing some copy paste along with your tests.

**Comment by tomschlick** [comment]
@JacobBennett Yeah sorry, I've been trying to find time over the past few days to get this rolling but other stuff got in the way. 

Probably wouldn't be able to get to it until this weekend so I'm perfectly ok with it if you want to put something together. No credit necessary :)

**Comment by ramsey** [comment]
I think @taylorotwell's suggestion is good. It seems to me that the crawler approach might end up sending resources to the client that shouldn't be sent yet. With an explicit approach, the developer has control over what they want (or don't want) to push.

**Comment by JacobBennett** [comment]
@ramsey do you have any opinions on having a cookie track previously pushed assets to prevent pushing them again? Is that a legit concern for push?

**Comment by ramsey** [comment]
I worry about tight coupling when trying to track pushed resources with a cookie.

In an ideal world, the browser (the browser itself and not the client-side, JavaScript application running in the browser) is keeping track of this and is telling the server (the web server—not the PHP application) to stop sending those resources it already has, and perhaps the web server itself (Nginx/Apache) is keeping track of this and not pushing those resources it knows it's already sent.

I'm not sure to what extent these components are already behaving in this way, though; I haven't done any research on that.

Right now, there's no good way to fully expose a web server's HTTP/2 functionality via the SAPI layer into PHP, so PHP applications can't effectively manage server push, which is why we're hacking it together with the preload link relationship.

Anyway, sorry for the digression, to answer your question: I would first look at how the client and server are behaving before attempting to track resources with a cookie. If the same resources are pushed by the server and received by the client on every request, then that's probably not an optimal use of preload server push.

Finally, I might implement the tracking on the server-side, rather than using a client-side approach; keep track of what resources have been pushed for each session and only push those that have not yet been pushed (or that have been updated in the meantime, if you can determine that).

**Comment by taylorotwell** [comment]
Will look over Tom's package but open to more ideas here if y'all want to take that approach with integrating something. I'll admit I'm not all that up to speed on this topic.
